netshare/generate/denormalize_fields.py

```diff
@@ -0,0 +1,111 @@
+import csv
+import os
+import random
+import tempfile
+from typing import List
+
+import numpy as np
+
+from netshare.generate import generate_api
+from netshare.learn import learn_api
+from netshare.utils.field import ContinuousField, Field
+
+
+def _get_fields_names(fields_list: List[Field]) -> List[str]:
+    """
+    This function returns the names of the given fields.
+    """
+    field_names = []
+    for field in fields_list:
+        if isinstance(field.name, list):
+            field_names.extend(field.name)
+        else:
+            field_names.append(field.name)
+    return field_names
+
+
+def _denormalize_by_fields_list(
+    normalized_data: np.ndarray, fields_list: List[Field], is_session_key: bool
+) -> List[np.ndarray]:
+    """
+    This function executes field.denormalize for each of the given field.
+    """
+    denormalized_data = []
+    dim = 0
+    for field in fields_list:
+        if is_session_key:
+            sub_data = normalized_data[:, dim : dim + field.getOutputDim()]
+        else:
+            sub_data = normalized_data[:, :, dim : dim + field.getOutputDim()]
+        sub_data = field.denormalize(sub_data)
+        if isinstance(field, ContinuousField):
+            sub_data = sub_data[:, 0] if is_session_key else sub_data[:, :, 0]
+        denormalized_data.append(sub_data)
+        dim += field.getOutputDim()
+    return denormalized_data
+
+
+def write_to_csv(
+    csv_folder: str,
+    session_key_fields: List[Field],
+    timeseries_fields: List[Field],
+    session_key: List[np.ndarray],
+    timeseries: List[np.ndarray],
+    data_gen_flag: np.ndarray,
+) -> None:
+    """
+    This function dumps the given data to the given directory as a csv format.
+    """
+    os.makedirs(csv_folder, exist_ok=True)
+    csv_path = os.path.join(csv_folder, f"data_{random.random()}.csv")
+    with open(csv_path, "w") as f:
+        writer = csv.writer(f)
+        session_titles = _get_fields_names(session_key_fields)
+        timeseries_titles = _get_fields_names(timeseries_fields)
+        writer.writerow(session_titles + timeseries_titles)
+
+        for i in range(session_key[0].shape[0]):
+            session_data = [d[i] for d in session_key]
+            # this if is here in parallel to the if in `reduce_samples`. It supports old flows.
+            if len(timeseries) == 1:
+                timeseries_data = timeseries[0][i].tolist()
+            else:
+                timeseries_data = [d[i][0] for d in timeseries]
+            writer.writerow(session_data + timeseries_data)
+
+
+def denormalize_fields() -> str:
```

Comment:
Timestamp may have some logical difference with other features in timeseries. For example, when the user is specifying "interarrival", we are calculating the delta of timestamp between two consecutive time steps in the same session as a fields. Therefore, when we are denormalizing the generated data, we should group by session key and also accumulate that field in order to get its absolute timestamp.

Reply:
hmm that's a really good point!
It seems that I missed that in my implementation, right?
can you refer my to the code in the old implementation, so I can double-check that I have everything?
If I understand correctly, the piece of code that you refer to is in `merge_attr` inside `model_generate`, and it's not here in the `denormalize_fields`.
If this is indeed what I think it is, and you think that it makes more sense, then I'll move it to here 👍

Reply:
I think it is previously in the function `denormalize`. After we finish generating, we need to merge all the data across chunks together and do `denormalization`. In the latest version of code in the branch new_dataset, we just commented out the `denormalization` since we haven't finished it for the generic dataset. We can also add it quickly.